### DISCUSSION TITLE
Seeding entries with existing relations intact?

### DISCUSSION THREAD

**Discussion by fleetwood** [prisma/prisma#11531] [Q&A]
Say I want to seed a number of countries:

```typescript
const entries: Prisma.CountryCreateInput[] = [
    {
        name: 'United States'
    },
    {
        name: 'Canada'
    },
    {
        name: 'UK'
    },
    {
        name: 'Germany'
    },
];
```

...and from there I want to seed users who are connected to those countries:

```typescript
const entries: Prisma.UserCreateInput[] = [
    {
        firstName: 'John',
        lastName: 'Danger',
        country:  {} // United States
    },
    {
        firstName: 'Annie',
        lastName: 'Anarchy',
        country:  {} // United States
    },
    {
        firstName: 'Charles',
        lastName: 'Profound',
        country:  {} // Germany
    },
    {
        firstName: 'Lucy',
        lastName: 'Blades',
        country:  {} // UK
    }
```

The seeds are organized into individual files in the directory, so I can't create them sequentially in the same file and then reference them.
- /seeds
-- Country.ts
-- Users.ts
-- index.ts

**The Models:**
```prisma
model User {
  id        Int        @id @default(autoincrement())
  firstName String
  lastName  String
  email     String       @unique
  countryId Int          // Required
  country   Country      @relation(fields: [countryId], references: [id]) // Required
}

model Country {
  id        Int        @id @default(autoincrement())
  name      String
  users User[]
}
```

How do I make those connections (relations) when running `prisma db seed`?

**Comment by austincrim** [comment]
Hey @fleetwood 👋 ,

It depends on how you have seeding set up right now. 

Could you provide more details about your `seed` command? 
Do `Country.ts` and `Users.ts` only contain data or do they also have seeding logic?

Thanks!

**Reply by fleetwood** [reply]
Friendly bump!

**Comment by fleetwood** [comment]
Each Model has its own set of data, and each follows a pseudo-abstraction:

**Country.ts**

```typescript
const entries: Prisma.CountryCreateInput[] = [
    {
        name: 'United States'
    },
    {
        name: 'Canada'
    },
    {
        name: 'UK'
    },
    {
        name: 'Germany'
    },
];

async function createCountries(prisma: PrismaClient) {
    console.log(`\tCreating ${label}...`);
    let success = 0;
    for (const data of entries) {
        await prisma.country.create({ data })
        success++;
    }
    console.log(`\tCreated ${label}: ${success} of ${entries.length}`);
}
```

**User.ts**
```typescript
const entries: Prisma.UserUncheckedCreateInput[] = [
    {
        firstName: 'John',
        lastName: 'Danger',
        email: 'john.danger@gmail.com',
        countryId: 2 // NOTE: hard-coding expected ids
    },
    ...
];

async function createUsers(prisma: PrismaClient) {
    console.log(`\tCreating ${label}...`);
    let success = 0;
    for (const data of entries) {
        await prisma.user.create({ data })
        success++;
    }
    console.log(`\tCreated ${label}s: ${success} of ${entries.length}`)
}
```

**seed.ts**:
```typescript
import createCountries from './seeds/Country';
import createUsers from './seeds/User';

const seeds = [
  createCountries,
  createUsers,
];

const prisma = new PrismaClient();

async function main() {
  console.log(`Start seeding ...`);
  seeds.forEach(seed => async () => await seed(prisma));
}

main()
  .catch((e) => {
    console.error(e)
    process.exit(1)
  })
  .finally(async () => {
    await prisma.$disconnect()
    console.log(`Seeding finished.`)
  })
```

The `create...` methods get executed asynchronously, each using the same passed `prisma` instance. I can guess at what the created ids might be and connect them that way, but obv that's a brittle approach.

**Reply by pantharshit00** [reply]
Hello @fleetwood, our apologies for the delay in getting back to you! Due to the large number of questions in our community channels, unfortunately this slipped through the cracks. It's very important for us to address your question, so let us know in a comment if this is still an open discussion. 

I would actually just store data in different files here and run the related queries in a single file. That way you can use nested creates and you will be able to properly seed the required data. Here is an example project which you can reference: https://github.com/freeCodeCamp/chapter/blob/main/server/prisma/seed/seed.ts

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
